### PR TITLE
[15.0] [MIG] partner_time_to_pay + improvements

### PR DESCRIPTION
**Pull request by Shide**
What i've done:
- All the things on [Migration Guidelines](https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-15.0#technical-method-to-migrate-a-module-from-140-to-150-branch) + marked as fixup the last black, isort, etc commit
- Add a new field `full_reconcile_payment_date` on invoices to show when has been full reconciled
- Make all the fields on partner store true and computed based on his invoices and the `full_reconcile_payment_date` field
- Refactor and simplify the number of functions (and the lines of code to maintain) to compute the partner time to pay, using the `full_reconcile_payment_date` field
- Views for invoices
- Refactor tests completely and remove unnecesary lines
- Renamed `res_partner.xml` to `res_partner_view.xml`
- Add `es.po` and the `partner_time_to_pay .pot`
- Improved README.md to show what has been added

MT-1494 @moduon

@Murtaza-OSI @rafaelbn @bosd @agyamuta @max3903 @yajo feel free to review and suggest anything

This PR can be closed https://github.com/OCA/account-invoice-reporting/pull/232 because it's not being maintained and this PR has a better performance (all stored true) and I'm going to fix and review as soon as possible anything it's wrong or you want to modify.

### CODE REVIEW

File: partner_time_to_pay/readme/DESCRIPTION.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+This module displays statistics related to the receivables and payables behavior of a partner on the Accounting tab of the partner form view.
+Also displays the full reconcile payment date on invoices
```
Comment (yajo) [inline partner_time_to_pay/readme/DESCRIPTION.rst]:
```suggestion
This module displays statistics related to the receivables and payables behavior of a partner on the *Sales & Purchase* tab of the partner form view.
Also displays the full reconcile payment date on invoices.
```

File: partner_time_to_pay/readme/USAGE.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+* Go to the Accounting tab of a partner
```
Comment (yajo) [inline partner_time_to_pay/readme/USAGE.rst]:
```suggestion
* Go to the *Sales & Purchase* tab of a partner
```

File: partner_time_to_pay/readme/USAGE.rst [hunk 3]
```diff
@@ -0,0 +1,4 @@
...
+* There is 4 new fields. These fields represent the average days to pay for
```
Comment (yajo) [inline partner_time_to_pay/readme/USAGE.rst]:
```suggestion
* There are 4 new fields. These fields represent the average days to pay for
```

File: partner_time_to_pay/readme/USAGE.rst [hunk 4]
```diff
@@ -0,0 +1,4 @@
...
+  this year (YTD) and since the first invoice (lifetime).
+* Adds a new field to invoices to check the Full Reconcile Payment Date
```
Comment (yajo) [inline partner_time_to_pay/readme/USAGE.rst]:
```suggestion
* Adds a new field in the invoices form to check the Full Reconcile Payment Date
```

File: partner_time_to_pay/models/res_partner.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    d2r_life = fields.Float(
+        compute="_compute_d2x",
+        string="AVG Days to Receivable (lifetime)",
+        store=True,
+    )
+    d2r_ytd = fields.Float(
+        compute="_compute_d2x",
+        string="AVG Days to Receivable (YTD)",
+        store=True,
+    )
```
Comment (yajo) [inline partner_time_to_pay/models/res_partner.py]: Please add `help` on these fields. The names use some acronyms that might not be obvious at first sight.

File: partner_time_to_pay/models/account_move.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        compute="_compute_full_reconcile_payment_date",
+        store=True,
+        help="Date when the complete reconciliation of this invoice occurred.",
+    )
+
+    @api.depends("payment_state")
+    def _compute_full_reconcile_payment_date(self):
+        in_payment_states = {self._get_invoice_in_payment_state()}
+        in_payment_states.add("paid")
+
```
Comment (yajo) [inline partner_time_to_pay/models/account_move.py]:
```suggestion
        in_payment_states = {"paid", self._get_invoice_in_payment_state()}
```

File: partner_time_to_pay/models/account_move.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        for move in self:
+            if move.payment_state in in_payment_states:
+                if not move.full_reconcile_payment_date:
+                    payments = move._get_reconciled_payments().sorted(
+                        "date", reverse=True
+                    )
+                    move.full_reconcile_payment_date = (
+                        payments and payments[0].date or fields.Date.today()
```
Comment (yajo) [inline partner_time_to_pay/models/account_move.py]: A little trick that does the same:

```suggestion
                        payments[:1].date or fields.Date.today()
```

**Review by yajo — CHANGES_REQUESTED** [review]
In general, it works fine.

In runboat I couldn't see invoices when logging in as `demo` user. I'm not sure if that's because of this module, or maybe it's another one producing the problem. Could yo please check that?

Some comments below.

**Comment by Shide** [comment]
> In general, it works fine.
> 
> In runboat I couldn't see invoices when logging in as `demo` user. I'm not sure if that's because of this module, or maybe it's another one producing the problem. Could yo please check that?
> 
> Some comments below.

Suggestions applied.
I've enter on another runboat from a month ago and when i'm entering as demo in invoices, happens as well

**Review by max3903 — APPROVED** [review]

File: partner_time_to_pay/README.rst [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+Do not contact contributors directly about support or help with technical issues.
+
+Credits
+=======
+
+Authors
+~~~~~~~
+
+* Open Source Integrators
+
```
Comment (rafaelbn) [inline partner_time_to_pay/README.rst:75]: I think the work is enough to be here, please add Moduon, thanks.

File: partner_time_to_pay/README.rst [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+Contributors
+~~~~~~~~~~~~
+
+* Ray Carnes <rcarnes@ursainfosystems.com>
+* Maxime Chambreuil <mchambreuil@opensourceintegrators.com>
+* Bhavesh Odedra <bodedra@opensourceintegrators.com>
+* Sandip Mangukiya <smangukiya@opensourceintegrators.com>
+* Serpent Consulting Services Pvt. Ltd. <support@serpentcs.com>
+* Jevin Dement <jdement@opensourceintegrators.com>
```
Comment (rafaelbn) [inline partner_time_to_pay/README.rst:84]: We have contribute too please Edu, both.

File: partner_time_to_pay/README.rst [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+   :alt: Odoo Community Association
+   :target: https://odoo-community.org
+
+OCA, or the Odoo Community Association, is a nonprofit organization whose
+mission is to support the collaborative development of Odoo features and
+promote its widespread use.
+
+.. |maintainer-max3903| image:: https://github.com/max3903.png?size=40px
+    :target: https://github.com/max3903
+    :alt: max3903
```
Comment (rafaelbn) [inline partner_time_to_pay/README.rst:101]: add me here please , this will informe me so I can test changes

**Review by rafaelbn — COMMENTED** [review]
Functionall test in runboat 👍 

Please review my comments before merging

imagen

File: partner_time_to_pay/__manifest__.py [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+{
+    "name": "Partner Time to Pay",
+    "summary": "Add receivables and payables statistics to partners",
+    "version": "15.0.1.0.0",
+    "license": "AGPL-3",
+    "development_status": "Beta",
+    "author": "Open Source Integrators, Odoo Community Association (OCA), Moduon",
+    "category": "Accounting & Finance",
+    "website": "https://github.com/OCA/account-invoice-reporting",
+    "maintainers": ["max3903"],
```
Comment (max3903) [inline partner_time_to_pay/__manifest__.py]:
```suggestion
    "maintainers": ["max3903", "rafaelbn"],
```

**Comment by Shide** [comment]
@rafaelbn @max3903 suggestions applied 😄

**Review by rafaelbn — APPROVED** [review]
👍

**Review by fcvalgar — APPROVED** [review]
Functional Review ok

**Pull request merged**